src/metadata/fam_metadata_manager_direct.cpp

```diff
@@ -915,20 +928,21 @@ int FAM_Metadata_Manager::Impl_::metadata_insert_dataitem(
                 int ret1 = dataitemNameKVS->Del(dataitemName.c_str(),
                                                 dataitemName.size());
                 if (ret1 != META_NO_ERROR) {
-                    return ret1;
+                    message << "Failed to remove dataitem name entry from KVS, "
+                               "while cleanup";
+                    throw Metadata_Service_Exception(METADATA_ERROR,
+                                                     message.str().c_str());
                 }
             }
-            return ret;
+            message << "Failed to insert metadata into dataitem Id KVS";
+            throw Metadata_Service_Exception(METADATA_ERROR,
+                                             message.str().c_str());
         }
-
-        return ret;
-
     } else {
-
-        DEBUG_STDERR(dataitemId, "Region lookup failed");
-        return META_ERROR;
+        DEBUG_STDOUT(regionName, "region not found");
+        message << "Regioin does not exist";
+        throw Metadata_Service_Exception(METADATA_ERROR, message.str().c_str());
```

Comment:
You might want to consider using new exception macros. It has added advantage of printing line number etc.